File: cpp/2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int digit=0;
        int remainder=0;
        ListNode *ans=new ListNode(0);
        ListNode *cur=ans;
        while(l1 || l2)
        {
            int x= (l1?l1->val :0);
            int y= (l2?l2->val :0);
            digit= (x+y+remainder)%10;
            remainder=(x+y+remainder)/10;
            cur->next=new ListNode(digit);
            cur=cur->next;
            if(l1)l1=l1->next;
            if(l2)l2=l2->next;
        }
        if(remainder)cur->next=new ListNode(remainder);
        return ans->next;
    }
};

File: golang/2.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    var ans *ListNode = new(ListNode)
    var pCur *ListNode = ans
    
    var digit1, digit2, carry, remain, sum int
    for {
        if l1 == nil || l2 == nil {
            break
        }
        digit1 = l1.Val
        digit2 = l2.Val
        sum = digit1 + digit2 + carry
        carry = sum / 10
        remain = sum % 10
        
        pCur.Next = &ListNode{remain, nil}
        pCur = pCur.Next
    
        l1 = l1.Next
        l2 = l2.Next
    }
    
    for {
        if l1 == nil {
            break
        }
        sum = l1.Val + carry
        carry = sum / 10
        remain = sum % 10
        
        pCur.Next = &ListNode{remain, nil}
        pCur = pCur.Next
    
        l1 = l1.Next
    }
    
    for {
        if l2 == nil {
            break
        }
        sum = l2.Val + carry
        carry = sum / 10
        remain = sum % 10
        
        pCur.Next = &ListNode{remain, nil}
        pCur = pCur.Next
    
        l2 = l2.Next
    }
    
    if carry > 0 {
        pCur.Next = &ListNode{carry, nil}
    }
    return ans.Next
}
